_posts/2021-02-26-dune-3-plan.md

```diff
@@ -0,0 +1,87 @@
+---
+layout: blog
+title: Plan for Dune 3.0
+author: andreymokhov
+tags: [dune]
+---
+
+<!-- TODO: After publishing the post, create a Discuss page, and link to it. -->
+
+The Dune team has started planning the next major release and we would like to
+share and discuss the current plan with the community. We expect to release
+Dune 3 in August 2021. As with previous major releases, we will put a lot of
+effort in backward compatibility, which means the majority of Dune users should
+not notice any changes.
+
+## What's new
+
+Here are the main features we plan include in Dune 3:
+
+* **Incremental file-watching mode.** Dune already supports a basic
+  file-watching mode, activated by the `-w` flag, where the build is restarted
+  whenever any source file changes. The current implementation is
+  non-incremental: restarting really means restarting the build from scratch,
+  which, for example, involves reparsing all `dune` files. On small-size
+  projects, this is fast enough, but this clearly doesn't scale well to large
+  monorepos. We are making the file-watching mode fully incremental, to avoid
+  redoing the computations whose inputs didn't change since the previous build.
+
+* **Dune RPC.** We are working on adding an RPC server to Dune that will make it
+  possible to control Dune programmatically (instead of using the command line),
+  which is often desirable for integration with various other tools, such as
+  editors.
+
+* **Better scoping semantics.** Dune provides some project-level scoping
+  functionality, for example, it's possible to mark a library as private to
+  prevent leaking the implementation details outside of a project. The current
+  implementation of scoping has wrong semantics for nested vendored projects: a
+  public library in the inner vendored project escapes its scope and is globally
+  visible. This loophole caused problems for users, and we are going to remove
+  it. This will be a breaking change but nested vendored projects are not very
+  common, so this change will only affect a small number of users and in a
+  positive way.
+
+* **Better support for warnings.** Currently, if Dune executes a build command
+  that reports a warning but doesn't fail, that warning will only be shown to
+  the user on the first build. Indeed, since the command succeeded, there is no
+  need to rerun it again in subsequent builds unless its inputs change, and so
+  the warning may go unnoticed and unfixed. This is why Dune turns warnings into
+  errors by default, which can be annoying. Dune users (and we too) often
+  complain about this, so we are working on better support for warnings.
+
+## Features we plan to remove
+
+We plan to remove the features listed below. Please let us know if you will be
+affected and will need some help with the transition to Dune 3.
+
+* **Support for OPAM v1.** We believe that it's not used any more, so we'd like
+  to remove the associated code.
+
```

Comment:
You can link to https://opam.ocaml.org/blog/deprecating-opam-1-2-0/ if useful -- it has been EOLed for some years now.

Reply:
Thanks for the link, added!